core/src/main/scala/kafka/server/AlterPartitionManager.scala

```diff
@@ -124,7 +125,20 @@ class DefaultAlterPartitionManager(
   val metadataVersionSupplier: () => MetadataVersion
 ) extends AlterPartitionManager with Logging with KafkaMetricsGroup {
 
-  // Used to allow only one pending ISR update per partition (visible for testing)
+  // Used to allow only one pending ISR update per partition (visible for testing).
+  // Note that we key items by TopicPartition despite using TopicIdPartition while
+  // submitting changes. We do this to ensure that topics with the same name but
+  // with a different topic id or no topic id collide here. There are two cases to
+  // consider:
+  // 1) When the cluster is upgraded from IBP < 2.8 to IBP >= 2.8, the ZK controller
+  //    assigns topic ids to the partitions. So partitions will start sending updates
+  //    with a topic id while they might still have updates without topic ids in this
+  //    Map. This would break the contract of only allowing one pending ISR update per
+  //    partition.
+  // 2) When a topic is deleted and re-created, we cannot have two entries in this Map
+  //    especially if we cannot use an AlterPartition request version which supports
+  //    topic ids in the end because the two updates with the same name would be merged
+  //    together.
```

Comment:
Okay. I think in most case this would fail because the partition epoch doesn't equal. It would only succeed if the controller created the new topic partition with the same assignment and leader, and the partition epoch was never increased for the old partition.